guides/keras_tuner/visualize_tuning.py

```diff
@@ -0,0 +1,180 @@
+the number of units or filters, whether to use dropout.
+"""
+
+
+def build_model(hp):
+    num_classes = 10
+    input_shape = (28, 28, 1)
+    model = keras.Sequential()
+    model.add(keras.layers.InputLayer(input_shape=input_shape))
+    # Model type can be MLP or CNN.
+    model_type = hp.Choice("model_type", ["mlp", "cnn"])
+    if model_type == "mlp":
+        model.add(layers.Flatten())
+        # Number of layers of the MLP is a hyperparameter.
+        for i in range(hp.Int("mlp_layers", 1, 3)):
+            # Number of units of each layer are
+            # different hyperparameters with different names.
+            model.add(
+                layers.Dense(
+                    units=hp.Int("units_{i}".format(i=i), 32, 128, step=32),
+                    activation="relu",
+                )
+            )
+    else:
+        # Number of layers of the CNN is also a hyperparameter.
+        for i in range(hp.Int("cnn_layers", 1, 3)):
+            model.add(
+                layers.Conv2D(
+                    hp.Int("filters_{i}".format(i=i), 32, 128, step=32),
+                    kernel_size=(3, 3),
+                    activation="relu",
+                )
+            )
+            model.add(layers.MaxPooling2D(pool_size=(2, 2)))
+        model.add(layers.Flatten())
+
+    # A hyperparamter for whether to use dropout layer.
+    if hp.Boolean("dropout"):
+        model.add(layers.Dropout(0.5))
+
+    model.add(layers.Dense(units=num_classes, activation="softmax"))
+
+    # Compile the model.
+    model.compile(
+        loss="sparse_categorical_crossentropy",
+        metrics=["accuracy"],
+        optimizer=hp.Choice("optimizer", ["adam", "adadelta", "sgd"]),
+    )
+    return model
+
+
+"""
+We can do a quick test of the models to check if it build successfully for both
+CNN and MLP.
+"""
+
+import keras_tuner as kt
+
+# Initialize the `HyperParameters` and set the values.
+hp = kt.HyperParameters()
+hp.values["model_type"] = "cnn"
+# Build the model using the `HyperParameters`.
+model = build_model(hp)
+# Test if the model runs with our data.
+model(x_train[:100])
+# Print a summary of the model.
+model.summary()
+
+# Do the same for MLP model.
+hp.values["model_type"] = "mlp"
+model = build_model(hp)
+model(x_train[:100])
+model.summary()
+
+"""
+Initialize the `RandomSearch` tuner with 10 trials and using validation
+accuracy as the metric for selecting models.
+"""
+
+tuner = kt.RandomSearch(
+    build_model,
+    max_trials=10,
+    # Do not resume the previous search in the same directory.
+    overwrite=True,
+    objective="val_accuracy",
+    # Set a directory to store the intermediate results.
+    directory="/tmp/tb",
+)
+
+"""
+Start the search by calling `tuner.search(...)`. To use TensorBoard, we need
+to pass a `keras.callbacks.TensorBoard` instance to the callbacks.
+"""
+
+tuner.search(
+    x_train,
+    y_train,
+    validation_data=(x_test, y_test),
```

Comment:
Never use the test set as validation data during tuning. In this example, you can use a `validation_split` and finally test your best model(s) on the test set at the end